File: algorithm/杭电ACM/ambition0109/1084.java
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: ambition0109
////Nickname: Ambition
////Run ID: 
////Submit time: 2011-04-11 20:39:12
////Compiler: Java
//////////////////////////////////////////////////////
////Problem ID: 1084
////Problem Title: 
////Run result: Accept
////Run time:109MS
////Run memory:7048KB
//////////////////System Comment End//////////////////
import java.util.Arrays;
import java.util.Scanner;

@SuppressWarnings("rawtypes")
class Node implements Comparable{
	String time;
	int score,pos;
	
	Node(String t,int i){
		this.time=t;
		this.pos=i;
	}
	
	public int compareTo(Object arg0){
		String t=((Node)arg0).time;
		return this.time.compareTo(t);
	}
	
	public String toString(){
		return (""+score);
	}
}

public class Main {
	public static void main(String args[]){
		Scanner in=new Scanner(System.in);
		Node[][] time=new Node[6][];
		int[] num=new int[6];
		int[] pos=new int[105];

		while(true){
			for(int i=0;i<6;i++){
				time[i]=new Node[105];
				num[i]=0;
			}
			int n=in.nextInt();
			if(n<0) break;
			for(int i=0;i<n;i++){
				int t=in.nextInt();
				String str=in.next();
				time[t][num[t]++]=new Node(str,i);
				pos[i]=t;
			}

			for(int i=0;i<num[0];i++) time[0][i].score=50;
			for(int i=0;i<num[5];i++) time[5][i].score=100;
			for(int i=1;i<5;i++){
				if(num[i]==0) continue;
				Arrays.sort(time[i],0,num[i]);
				int j=0;
				for(j=0;j<num[i]/2;j++){
					time[i][j].score=i*10+55;
				}
				for(;j<num[i];j++){
					time[i][j].score=i*10+50;
				}
			}
			for(int i=0;i<n;i++){
				for(int j=0;j<num[pos[i]];j++){
					if(time[pos[i]][j].pos==i){
						System.out.println(time[pos[i]][j]);
					}
				}
			}
			System.out.println();
		}
	}
}

File: algorithm/杭电ACM/杭电ACM_C++/杭电题/1084.cpp
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: devil_momo
////Nickname: Lucifer
////Run ID: 
////Submit time: 2010-03-21 22:56:51
////Compiler: GUN C++
//////////////////////////////////////////////////////
////Problem ID: 1084
////Problem Title: 
////Run result: Accept
////Run time:0MS
////Run memory:216KB
//////////////////System Comment End//////////////////
#include <algorithm>
using namespace std;
struct REC
{
	int i,p,t;
}rec[101];
int score[6];
int final[101];
int trans[5]={50,60,70,80,90};
bool cmp(REC a,REC b)
{
	if(a.p!=b.p)
		return a.p>b.p;
	else
		return a.t<b.t;
}
int main()
{
	int n,hh,mm,ss;
	int i,j,k;
	while(scanf("%d",&n),n>=0)
	{
		memset(score,0,sizeof(score));
		memset(rec,0,sizeof(rec));
		for(i=0;i<n;i++)
		{
			scanf("%d %d:%d:%d",&rec[i].p,&hh,&mm,&ss);
			rec[i].i=i;
			rec[i].t=hh*10000+mm*100+ss;
			score[rec[i].p]++;
		}
		sort(rec,rec+n,cmp);
		k=0;
		for(j=0;j<score[5];j++)
			final[rec[k++].i]=100;
		for(i=4;i>0;i--)
		{
			for(j=0;j<score[i]/2;j++)
			{
				final[rec[k++].i]=trans[i]+5;
			}
			for(;j<score[i];j++)
			{
				final[rec[k++].i]=trans[i];
			}
		}
		while(k<n)
			final[rec[k++].i]=50;
		for(i=0;i<n;i++)
			printf("%d\n",final[i]);
		puts("");
	}
	return 0;
}